libs/src/services/creatorNode/CreatorNodeSelection.js

```diff
@@ -0,0 +1,249 @@
+        return services.map(e => e.endpoint)
+      },
+      whitelist
+    })
+    this.creatorNode = creatorNode
+    this.numberOfNodes = numberOfNodes
+    this.ethContracts = ethContracts
+    this.blacklist = blacklist
+    // Services with the structure {request, response, millis} (see timeRequest) sorted by semver and response time
+    this.healthCheckedServicesList = []
+
+    // List of valid past creator node versions registered on chain
+    this.validVersions = null
+  }
+
+  /**
+   * Selects a primary and secondary creator nodes
+   *
+   * 1. Retrieve all the creator node services
+   * 2. Filter from/out creator nodes based off of the whitelist and blacklist
+   * 3. Filter out unhealthy, outdated, and still syncing nodes via health and sync check
+   * 4. Sort by healthiest (highest version -> lowest version); secondary check if equal version based off of responseTime)
+   * 5. Select a primary and numberOfNodes-1 number of secondaries (most likely 2) either from available nodes or backups
+   */
+  async select () {
+    // Reset decision tree and backups
+    this.decisionTree = []
+    this.clearBackups()
+
+    // Get all the creator node endpoints on chain and filter
+    let services = await this.getServices()
+    this.decisionTree.push({ stage: 'Get All Services', val: services })
+
+    if (this.whitelist && this.whitelist.length > 0) { services = this.filterToWhitelist(services) }
+    this.decisionTree.push({ stage: 'Filtered To Whitelist', val: services })
+
+    if (this.blacklist && this.blacklist.length > 0) { services = this.filterFromBlacklist(services) }
+    this.decisionTree.push({ stage: 'Filtered Out From Blacklist', val: services })
+
+    const healthCheckedServices = await this.timeRequests(
+      services.map(node => ({
+        id: node,
+        url: `${node}/health_check`
+      }))
+    )
+    // Store a copy of the sorted by version and resposne time nodes
+    this.healthCheckedServicesList = healthCheckedServices
+
+    const healthyServices = healthCheckedServices.filter(resp => {
+      const endpoint = resp.request.id
+
+      // Add everything as backup regardless of response
+      this.addBackup(endpoint, resp.response)
+
+      let isHealthy = false
+      if (resp.response) {
+        const isUp = resp.response.status === 200
+        const versionIsUpToDate = this.ethContracts.hasSameMajorAndMinorVersion(this.currentVersion, resp.response.data.data.version)
+        isHealthy = isUp && versionIsUpToDate
+      }
+
+      // If not healthy, remove as backup and add to unhealthy
+      if (!isHealthy) {
+        this.removeFromBackups(endpoint)
+        this.addUnhealthy(endpoint)
+      }
+      return isHealthy
+    })
+    services = healthyServices.map(service => service.request.id)
+
+    this.decisionTree.push({ stage: 'Filtered Out Known Unhealthy And Outdated Nodes', val: services })
+
+    const successfulSyncCheckServices = []
+    for (const service of services) {
```

Comment:
this loop should be done in parallel rather than awaiting inside each iteration

services.forEach(async service => { ... })